enterprise/server/remote_execution/containers/docker/docker.go

```diff
@@ -83,6 +91,34 @@ func (r *dockerCommandContainer) Run(ctx context.Context, command *repb.Command,
 		return result
 	}
 
+	containerRootDir := workDir
+	if r.options.EnableCASFS && r.env.GetConfigurator().GetExecutorConfig().EnableCASFS {
+		casfsDir := workDir + "_casfs"
+		if err := os.Mkdir(casfsDir, 0755); err != nil {
+			result.Error = status.UnavailableErrorf("could not create FUSE FS dir: %s", err)
+			return result
+		}
+
+		cfs := casfs.New(workDir, &casfs.Options{})
+		if err := cfs.Mount(casfsDir); err != nil {
+			result.Error = status.UnavailableErrorf("unable to mount CASFS at %q: %s", casfsDir, err)
+			return result
+		}
+		fileFetcher := dirtools.NewBatchFileFetcher(ctx, task.GetExecuteRequest().GetInstanceName(), r.env.GetFileCache(), r.env.GetByteStreamClient(), r.env.GetContentAddressableStorageClient())
+		err = cfs.PrepareForTask(ctx, fileFetcher, task.GetExecutionId(), fsLayout)
+		if err != nil {
+			result.Error = status.UnavailableErrorf("unable to prepare CASFS layout at %q: %s", casfsDir, err)
+			return result
+		}
+		defer func() {
+			err := cfs.Unmount()
+			if err != nil {
+				log.Warningf("CASFS unmount failed: %s", err)
+			}
+		}()
+		containerRootDir = casfsDir
+	}
```

Comment:
Am I understanding this right -- the only change to the container impl is that we're just swapping out the implementation of the `workingDir` from a regular directory to a FUSE-based one?

If so it seems like this logic should done at the Runner level (rather than the container level) and the container implementations don't need to care about how the `workingDir` is implemented -- otherwise you have to do this same logic for every container impl.

Reply:
For Firecracker, the FUSE filesystem needs to be mounted within the Guest VM so a generalized approach won't work for all container impls.

Reply:
I think maybe it makes sense then to do something like this in `runner.go`:

```
if fuseEnabled && !firecracker {
  workspaceDir = createFUSEWorkspace(ctx, env, workspaceDir, ...)
}
```

Because other container impls (bare, containerd, podman etc.) would all use the same logic.

Trying to avoid adding business logic to the container implementations as much as possible, since it's simpler to understand and gives us more flexibility if the container interface maps more or less directly to the equivalent CLI command.

Reply:
Done.